cpp/src/arrow/compute/kernels/scalar_hash.cc

```diff
@@ -0,0 +1,218 @@
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <algorithm>
+
+#include "arrow/array/array_base.h"
+#include "arrow/array/builder_primitive.h"
+#include "arrow/compute/kernels/common_internal.h"
+#include "arrow/compute/key_hash_internal.h"
+#include "arrow/compute/light_array_internal.h"
+#include "arrow/compute/util.h"
+#include "arrow/result.h"
+
+namespace arrow {
+namespace compute {
+namespace internal {
+
+// Define symbols visible within `arrow::compute::internal` in this file;
+// these symbols are not visible outside of this file.
+namespace {
+
+// ------------------------------
+// Kernel implementations
+// It is expected that HashArrowType is either UInt32Type or UInt64Type (default)
+
+template <typename ArrowType, typename Hasher>
+struct FastHashScalar {
+  using c_type = typename ArrowType::c_type;
+
+  static Result<KeyColumnArray> ToColumnArray(
+      const ArraySpan& array, LightContext* ctx,
+      const uint8_t* list_values_buffer = nullptr) {
+    KeyColumnMetadata metadata;
+    const uint8_t* validity_buffer = nullptr;
+    const uint8_t* fixed_length_buffer = nullptr;
+    const uint8_t* var_length_buffer = nullptr;
+
+    if (array.GetBuffer(0) != nullptr) {
+      validity_buffer = array.GetBuffer(0)->data();
+    }
+    if (array.GetBuffer(1) != nullptr) {
+      fixed_length_buffer = array.GetBuffer(1)->data();
+    }
+
+    auto type = array.type;
+    auto type_id = type->id();
+    if (type_id == Type::NA) {
+      metadata = KeyColumnMetadata(true, 0, true);
+    } else if (type_id == Type::BOOL) {
+      metadata = KeyColumnMetadata(true, 0);
+    } else if (is_fixed_width(type_id)) {
+      metadata = KeyColumnMetadata(true, type->bit_width() / 8);
+    } else if (is_binary_like(type_id)) {
+      metadata = KeyColumnMetadata(false, sizeof(uint32_t));
+      var_length_buffer = array.GetBuffer(2)->data();
+    } else if (is_large_binary_like(type_id)) {
+      metadata = KeyColumnMetadata(false, sizeof(uint64_t));
+      var_length_buffer = array.GetBuffer(2)->data();
+    } else if (type_id == Type::MAP) {
+      metadata = KeyColumnMetadata(false, sizeof(uint32_t));
+      var_length_buffer = list_values_buffer;
+    } else if (type_id == Type::LIST) {
+      metadata = KeyColumnMetadata(false, sizeof(uint32_t));
+      var_length_buffer = list_values_buffer;
+    } else if (type_id == Type::LARGE_LIST) {
+      metadata = KeyColumnMetadata(false, sizeof(uint64_t));
+      var_length_buffer = list_values_buffer;
+    } else if (type_id == Type::FIXED_SIZE_LIST) {
+      auto list_type = checked_cast<const FixedSizeListType*>(type);
+      metadata = KeyColumnMetadata(true, list_type->list_size());
+      fixed_length_buffer = list_values_buffer;
+    } else {
+      return Status::TypeError("Unsupported column data type ", type->name(),
```

Comment:
Continue my other comment in `compute.rst`, if we keep `Any` in the `Input types`, then will it be more appropriate to throw `NotImplemented` instead of `TypeError`?

Reply:
Instead, we could probably use a dedicated `TypeMatcher` implementation?

Reply:
Added a type matcher explicitly disallowing the union/view/REE types.